### ISSUE TITLE
Incorrect overload resolution for constrained self type in Generic class

### ISSUE THREAD

**Issue by CaselIT** [python/mypy#20765]
**Bug Report**

Mypy fails to resolve a constrained generic when applied to self, it works fine if it's on an argument

**To Reproduce**

https://mypy-play.net/?mypy=latest&python=3.14&gist=2b566dc3fa9787cf75b7517168ad2ff7

```py
from typing import Any, Generic, TypeVar, assert_type, overload
class A:
    pass

class C(A):
    pass

_a = TypeVar("_a", bound=A)
_a1 = TypeVar("_a1", bound=A)
_c = TypeVar("_c", bound=C)

class G(Generic[_a]):
    x: _a

    @overload
    def go(self: "G[_c]") -> int: ...
    @overload  # using _a is the same
    def go(self: "G[_a1]") -> str | int: ...

    def go(self) -> int | str:
        return 1 if isinstance(self.x, C) else "a"

    @overload
    def og(self, a: "G[_c]") -> int: ...
    @overload  # using _a is the same
    def og(self, a: "G[_a1]") -> str | int: ...

    def og(self, a: "G[Any]") -> int | str:
        return 1 if isinstance(self.x, C) else "a"

def a(x: G[A], y: G[C]) -> None:
    assert_type(x.go(), str | int)
    assert_type(y.go(), int)
    assert_type(x.og(x), str | int)
    assert_type(y.og(y), int)
```

**Expected Behavior**

No error should be reported, but mypy thinks `G[A]` is `G[C]` if the constraint is on self, works fine if it's an argument instead

**Actual Behavior**

```
main.py:37: error: Expression is of type "int", not "str | int"  [assert-type]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 1.19.1
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files):
- Python version used: 3.13 / 3.14

**Comment by ilevkivskyi** [comment]
I think this is essentially the same as https://github.com/python/mypy/issues/14243, we don't check type variable upper bounds when applying self-types

**Comment by CaselIT** [comment]
it does seem similar, I could not find it while searching. I guess I'll have to live with it.

Will add this example to the other issue
